File: Data Structures & Algorithms/binary-search/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int i = 0;
        int j = nums.size() - 1;
        int middle = j + ((i - j) / 2);
        int index = -1;

        while(i <= j){
            if(target > nums[middle]){
                i = middle + 1;
            }
            else if(target < nums[middle]){
                j = middle - 1;
            }
            else{
                return middle;
            }
            middle = j + ((i - j) / 2);
        }

        return index;

    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) {

        nums.sort((a, b) => a-b);
        let results = [];
  
        for(let i = 0; i<nums.length; i++){
            if(i > 0 && nums[i] === nums[i-1]) continue;
            let res = this.twoSum(-nums[i], nums.slice(i+1));
            if(res.length !== 0) results.push(...res);
        }

        return results;

    }

    twoSum(target, nums){
        let i = 0;
        let j = nums.length - 1;
        let results = [];
        while(i < j){
            let res = [];
            let sum = nums[i] + nums[j];
            if(sum > target) j--;
            else if(sum < target) i++;
            else{res.push([-target, nums[i], nums[j]]); results.push(res); i++; j--;};
            while (i < j && nums[i] === nums[i - 1]) i++;
while (i < j && nums[j] === nums[j + 1]) j--;

        }
        return results;
    }

}

// -6 -4 -2 -1 0 1 4 5 6 

// -4 -1 -1 0 1 2 

